validator/proposer/service.go

```diff
@@ -159,43 +161,121 @@ func (p *Proposer) run(done <-chan struct{}, client pb.ProposerServiceClient) {
-			// the proposal.
-			if latestBeaconBlock == nil {
-				log.Errorf("Could not marshal nil latest beacon block")
-				continue
-			}
-			data, err := proto.Marshal(latestBeaconBlock)
-			if err != nil {
-				log.Errorf("Could not marshal latest beacon block: %v", err)
-				continue
-			}
-			latestBlockHash := hashutil.Hash(data)
-
-			// To prevent any unaccounted attestations from being added.
-			p.lock.Lock()
-
-			bitmask := p.GenerateBitmask(p.pendingAttestation)
-			// TODO(#619): Implement real proposals with randao reveals and attestation fields.
-			req := &pb.ProposeRequest{
-				ParentHash: latestBlockHash[:],
-				// TODO(#511): Fix to be the actual, timebased slot number instead.
-				SlotNumber:         latestBeaconBlock.GetSlot() + 1,
-				RandaoRevealHash32: []byte{},
-				AttestationBitmask: bitmask,
-				Timestamp:          ptypes.TimestampNow(),
-			}
-			res, err := client.ProposeBlock(p.ctx, req)
-			if err != nil {
-				log.Errorf("Could not propose block: %v", err)
-				continue
-			}
+func (p *Proposer) receiveAssignment(latestBeaconBlock *pbp2p.BeaconBlock, client pb.ProposerServiceClient) {
+	log.Info("Performing proposer responsibility")
 
-			log.Infof("Block proposed successfully with hash %#x", res.BlockHash)
-			p.pendingAttestation = nil
-			p.lock.Unlock()
-		}
+	block, err := p.computeBlockToBeProposed(latestBeaconBlock, client)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	proposalData, err := p.createProposalDataFromBlock(block)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	signature := p.signProposalData(proposalData)
+	block.Signature[0] = signature[:]
+
+	hash, err := client.ProposeBlock(p.ctx, block)
+	if err != nil {
+		log.Errorf("Unable to propose block %v", err)
+		return
+	}
+
+	log.Infof("Successfully proposed block with hash %#x", hash.BlockHash)
+}
+
+func (p *Proposer) computeBlockToBeProposed(latestBlock *pbp2p.BeaconBlock,
+	client pb.ProposerServiceClient) (*pbp2p.BeaconBlock, error) {
+	// Extract the hash of the latest beacon block to use as parent hash in
+	// the proposal.
+	if latestBlock == nil {
+		return nil, fmt.Errorf("could not marshal nil latest beacon block")
+
+	}
+	data, err := proto.Marshal(latestBlock)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal latest beacon block: %v", err)
+	}
+	latestBlockHash := hashutil.Hash(data)
+
+	// Error is thrown away here because the client method returns no possible error.
+	// #nosec G104
+	powChainHashRes, _ := client.CurrentPOWChainBlockHash(p.ctx, nil)
+
+	indexRes, err := client.ProposerIndex(p.ctx, &pb.ProposerIndexRequest{SlotNumber: latestBlock.Slot})
+	if err != nil {
+		return nil, fmt.Errorf("could not get proposer index: %v", err)
 	}
+
+	proposerBitfield := uint64(math.Pow(2, 7-float64(indexRes.Index)))
+	attestation := &pbp2p.Attestation{
+		ParticipationBitfield: []byte{byte(proposerBitfield)},
+	}
+
+	// To prevent any unaccounted attestations from being added.
+	p.lock.Lock()
```

Comment:
Please call defer `p.lock.Unlock()` immediately after obtaining the lock. There is a code path where this method returns without unlocking the resource.